[RISCV][ISel] Use vaaddu with rounding mode rdn for ISD::AVGFLOORU.

sun-jacobi opened the pull request:
This patch aims to use `vaaddu` with rounding mode rdn (i.e `vxrm[1:0] = 0b10`) for `ISD::AVGFLOORU`.


### Source code 
```
define <8 x i8> @vaaddu_auto(ptr %x, ptr %y, ptr %z) {
  %xv = load <8 x i8>, ptr %x, align 2
  %yv = load <8 x i8>, ptr %y, align 2
  %xzv = zext <8 x i8> %xv to <8 x i16>
  %yzv = zext <8 x i8> %yv to <8 x i16>
  %add = add nuw nsw <8 x i16> %xzv, %yzv
  %div = lshr <8 x i16> %add, <i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1>
  %ret = trunc <8 x i16> %div to <8 x i8>
  ret <8 x i8> %ret 
}
```

### Before this patch 
```
vaaddu_auto: 
        vsetivli        zero, 8, e8, mf2, ta, ma
        vle8.v  v8, (a0)
        vle8.v  v9, (a1)
        vwaddu.vv       v10, v8, v9
        vnsrl.wi        v8, v10, 1
        ret
```
### After this patch 
```
vaaddu_auto: 
	vsetivli	zero, 8, e8, mf2, ta, ma
	vle8.v	v8, (a0)
	vle8.v	v9, (a1)
	csrwi	vxrm, 2
	vaaddu.vv	v8, v8, v9
	ret
```

### Note on signed averaging addition

Based on the rvv spec,  there is also a variant for signed averaging addition called `vaadd`.   
But AFAIU, no matter in which rounding mode, we cannot achieve the semantic of signed averaging addition through `vaadd`.   
Thus this patch only introduces `vaaddu`

topperc commented:
I don't think this is the right way to do this. We should make ISD::AVGFLOORU Custom for RISC-V and let DAGCombiner create an ISD::AVGFLOORU node that we can custom lower to a new RISCVISD::VAADDU_VL opcode.

sun-jacobi commented:
> I don't think this is the right way to do this. We should make ISD::AVGFLOORU Custom for RISC-V and let DAGCombiner create an ISD::AVGFLOORU node that we can custom lower to a new RISCVISD::VAADDU_VL opcode.

Thank you for the advice, I will try this way.

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -12357,6 +12393,51 @@ static SDValue combineAddOfBooleanXor(SDNode *N, SelectionDAG &DAG) {
+  return DAG.getNode(ISD::AVGFLOORU, DL, VT, N0->getOperand(0),
+                     N1->getOperand(0));
+}
+
```
Do we not already get ISD::AVGFLOORU nodes from `combineShiftToAVG` in TargetLowering.cpp?

lukel97 commented on llvm/test/CodeGen/RISCV/rvv/vaadd-autogen.ll:
```diff
@@ -0,0 +1,336 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc -mtriple=riscv32 -mattr=+d,+zfh,+zvfh,+v -target-abi=ilp32d \
+; RUN:   -verify-machineinstrs < %s | FileCheck %s
+; RUN: llc -mtriple=riscv64 -mattr=+d,+zfh,+zvfh,+v -target-abi=lp64d \
+; RUN:   -verify-machineinstrs < %s | FileCheck %s
```
Nit, I think you can just simplify the invocations to

```suggestion
; RUN: llc -mtriple=riscv32 -mattr=+v -verify-machineinstrs < %s | FileCheck %s
; RUN: llc -mtriple=riscv64 -mattr=+v -verify-machineinstrs < %s | FileCheck %s
```

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -1375,7 +1379,8 @@ RISCVTargetLowering::RISCVTargetLowering(const TargetMachine &TM,
                        ISD::INTRINSIC_WO_CHAIN, ISD::ADD, ISD::SUB, ISD::AND,
-                       ISD::OR, ISD::XOR, ISD::SETCC, ISD::SELECT});
+                       ISD::TRUNCATE, ISD::OR, ISD::XOR, ISD::SETCC,
+                       ISD::SELECT});
```
We don't need this since we don't have the combine anymore

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -1385,9 +1390,6 @@ RISCVTargetLowering::RISCVTargetLowering(const TargetMachine &TM,
   if (Subtarget.hasStdExtZbb())
     setTargetDAGCombine({ISD::UMAX, ISD::UMIN, ISD::SMAX, ISD::SMIN});
 
-  if (Subtarget.hasStdExtZbs() && Subtarget.is64Bit())
-    setTargetDAGCombine(ISD::TRUNCATE);
-
```
We can restore this too

lukel97 commented on llvm/test/CodeGen/RISCV/rvv/vaaddu-autogen.ll:
Could we rename this to `vaaddu-sdnode.ll` to match the other tests? I believe we usually use `vfoo.ll` for intrinsic tests and `vfoo-sdnode.ll` for tests on generic SD nodes. 

We also typically just use scalable vectors since they're the "native" vector type. Fixed vector tests then go into `fixed-vectors-vfoo.ll`.

lukel97 commented on llvm/test/CodeGen/RISCV/rvv/vaaddu-sdnode.ll:
Can we add some tests for vaaddu.vx? Something like

```llvm
define <vscale x 8 x i8> @vaaddu_vx_nxv8i8(<vscale x 8 x i8> %x, i8 %y) {
  %xzv = zext <vscale x 8 x i8> %x to <vscale x 8 x i16>
  %yhead = insertelement <vscale x 8 x i8> poison, i8 %y, i32 0
  %ysplat = shufflevector <vscale x 8 x i8> %yhead, <vscale x 8 x i8> poison, <vscale x 8 x i32> zeroinitializer
  %yzv = zext <vscale x 8 x i8> %ysplat to <vscale x 8 x i16>
  ...
}
```

lukel97 commented on llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td:
```diff
@@ -115,6 +124,7 @@ def riscv_saddsat_vl   : SDNode<"RISCVISD::SADDSAT_VL", SDT_RISCVIntBinOp_VL, [S
 def riscv_uaddsat_vl   : SDNode<"RISCVISD::UADDSAT_VL", SDT_RISCVIntBinOp_VL, [SDNPCommutative]>;
 def riscv_ssubsat_vl   : SDNode<"RISCVISD::SSUBSAT_VL", SDT_RISCVIntBinOp_VL>;
 def riscv_usubsat_vl   : SDNode<"RISCVISD::USUBSAT_VL", SDT_RISCVIntBinOp_VL>;
+def riscv_uavgadd_vl   : SDNode<"RISCVISD::UAVGADD_VL", SDT_RISCVIntBinOp_RM_VL, [SDNPCommutative]>;
```
I think most of the _vl nodes try to correspond to an SDNode in ISDOpcodes.h. Would it make more sense to instead:

- Mark `ISD::AVGFLOORU` as legal for scalable vector types, and custom for fixed length vector types
- Add a `RISCVISD::AVGFLOORU_VL` node with no rounding mode operand, i.e. it uses`SDT_RISCVIntBinOp_VL`
- Map `ISD::AVGFLOORU` to `RISCVISD::AVGFLOORU_VL` in `getRISCVVLOp`
- In `RISCVTargetLowering::LowerOperation`, add a case for `ISD::AVGFLOORU` beside `ISD::SADDSAT` and friends so fixed length vectors gets lowered via `lowerToScalableOp`
- Add a pattern for `avgflooru` in RISCVInstrInfoVSDPatterns.td, and `riscv_avgflooru_vl` in RISCVInstrInfoVVLPatterns.td, where both select `0b10` for the rounding mode in the pattern output

That way we wouldn't need the custom lowering in `RISCVTargetLowering::lowerUnsignedAvgFloor`, and it would follow the same codegen path as the saturating add and subtract instructions.

sun-jacobi replied:
Thank for your detailed advice. I will try this way.

lukel97 reviewed (approved):
Just a couple of nits but otherwise LGTM, thanks! Should probably wait for a second approval too in case anyone has other opinions

wangpc-pp reviewed (approved):
LGTM.

topperc reviewed (approved):
LGTM

sun-jacobi committed:
[RISCV][ISel] Use vaaddu with rounding mode rdn for ISD::AVGFLOORU.

This patch aims to use `vaaddu` with rounding mode rdn (i.e `vxrm[1:0] = 0b10`) for `ISD::AVGFLOORU`.

Source code
```
define <8 x i8> @vaaddu_auto(ptr %x, ptr %y, ptr %z) {
  %xv = load <8 x i8>, ptr %x, align 2
  %yv = load <8 x i8>, ptr %y, align 2
  %xzv = zext <8 x i8> %xv to <8 x i16>
  %yzv = zext <8 x i8> %yv to <8 x i16>
  %add = add nuw nsw <8 x i16> %xzv, %yzv
  %div = lshr <8 x i16> %add, <i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1>
  %ret = trunc <8 x i16> %div to <8 x i8>
  ret <8 x i8> %ret
}
```
Before this patch
```
vaaddu_auto:
        vsetivli        zero, 8, e8, mf2, ta, ma
        vle8.v  v8, (a0)
        vle8.v  v9, (a1)
        vwaddu.vv       v10, v8, v9
        vnsrl.wi        v8, v10, 1
        ret
```
After this patch
```
vaaddu_auto:
	vsetivli	zero, 8, e8, mf2, ta, ma
	vle8.v	v8, (a0)
	vle8.v	v9, (a1)
	csrwi	vxrm, 2
	vaaddu.vv	v8, v8, v9
	ret
```

Note on signed averaging addition

Based on the rvv spec,  there is also a variant for signed averaging addition called `vaadd`.
But AFAIU, no matter in which rounding mode, we cannot achieve the semantic of signed averaging addition through `vaadd`. Thus this patch only introduces `vaaddu`.
